### PR TITLE
Add auto-configuration for Spring Data Envers

### LINKED ISSUE #21370: Auto configuration for Spring Data Envers
When using [Spring Data Envers](https://github.com/spring-projects/spring-data-envers) with Spring Boot, the `EnversRevisionRepositoryFactoryBean` has to be configured via `@EnableJpaRepositories`:
```java
@EnableJpaRepositories(repositoryFactoryBeanClass = EnversRevisionRepositoryFactoryBean.class)
```

Maybe an auto configuration could take care of it. Would it make sense?

In case it does, I would be happy to prepare a pull request.

### PR DESCRIPTION
**Pull request by scordio**
This adds the auto-configuration for [Spring Data Envers](https://github.com/spring-projects/spring-data-envers).

Fixes #21370.

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java [hunk 1]
```diff
@@ -66,10 +69,23 @@
 @ConditionalOnMissingBean({ JpaRepositoryFactoryBean.class, JpaRepositoryConfigExtension.class })
 @ConditionalOnProperty(prefix = "spring.data.jpa.repositories", name = "enabled", havingValue = "true",
 		matchIfMissing = true)
-@Import(JpaRepositoriesRegistrar.class)
 @AutoConfigureAfter({ HibernateJpaAutoConfiguration.class, TaskExecutionAutoConfiguration.class })
 public class JpaRepositoriesAutoConfiguration {
 
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnClass(EnversRevisionRepositoryFactoryBean.class)
```
Comment (scordio) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java:76]: Is this a good target for the conditional? Mostly a style question.

I would have favored `org.springframework.data.envers.repository.support.EnversRevisionRepository` but it is deprecated. Another option could be `org.springframework.data.envers.repository.support.EnversRevisionRepositoryImpl`.

Comment (scordio) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java:76]: This conditional causes all the Spring Boot tests to use the Spring Data Envers auto-configuration implicitly as the dependency is always in the classpath (except when the `FilteredClassLoader` is used).

Is it desired or can it be an issue? Any suggestion to have something better?

**Comment by scordio** [comment]
Another open point is how to address testing properly. My preference would have been something like:

```java
assertThat(context).getBean(CityRepository.class)
	.extracting(AopProxyUtils::getSingletonTarget)
	.isInstanceOf(EnversRevisionRepositoryImpl.class); // doesn't work, it's always SimpleJpaRepository.class
```

but it doesn't work as the bean is always a `SimpleJpaRepository.class` although with the debugger I can see `EnversRevisionRepositoryImpl` being used. Any suggestion for a proper assertion?

Otherwise, I could add a `CityRevisionRepository` and perform verification on that, for both cases when Spring Data Envers is in the classpath or it is not.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java [hunk 2]
```diff
@@ -109,4 +111,18 @@ private AsyncTaskExecutor determineBootstrapExecutor(Map<String, AsyncTaskExecut
 
 	}
 
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnClass(EnversRevisionRepositoryFactoryBean.class)
+	@Import(EnversRevisionRepositoriesRegistrar.class)
```
[the remark this answers is not in the document]
Comment (wilkinsona) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java:116]: LGTM. `EnversRevisionRepository` would have been my first choice too. `EnversRevisionRepositoryFactoryBean` seems like a good alternative to me.

[the remark this answers is not in the document]
Comment (wilkinsona) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java:116]: That's not an uncommon problem due to `spring-boot-autoconfigure` being a bit monolithic. It should be fine but it is worth thinking about a bit. Thanks for raising it.

[the remark this answers is not in the document]
Comment (wilkinsona) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java:116]: Having thought so more, the scope of this is pretty narrow as most of our auto-configuration tests use `ApplicationContextRunner` where we choose which auto-configurations to be included. There are a handful of tests that use everything, and those being affected should not be a problem.

[the remark this answers is not in the document]
Comment (scordio) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/jpa/JpaRepositoriesAutoConfiguration.java:116]: Thanks for checking!

**Comment by scordio** [comment]
@wilkinsona sorry for the slow feedback, the PR should now be ready for review. I am wondering if some more test cases could be added, I will continue thinking about that.

I see the build is failing but I guess not because of my changes. I will keep an eye on `master` to see if a rebase if needed.

**Comment by scordio** [comment]
I finished to refactor the tests the way I had in mind and I don't plan to have any other change. Please let me know if there is something to be adjusted.

**Comment by scordio** [comment]
Hi everyone, is there anything else I could do to move forward?

**Comment by snicoll** [comment]
@scordio thank you for your patience and sorry we didn't have the time to get back to you yet. We're busy handling the scheduled back for `2.4.x` so it may take a bit more time before we can spend time on this one.

**Comment by scordio** [comment]
@snicoll thanks for the feedback and no problem, it was just a friendly reminder 🙂

**Comment by scordio** [comment]
Hi @snicoll, just wanted to check if this is blocked for any reason and if I can help with it.

**Comment by snicoll** [comment]
Hello @scordio thank you for your patience. I did review the proposal during the holidays and wanted to experiment how we could avoid having to subclass `JpaRepositoriesRegistrar`.

In retrospect, I should have written that down to get some feedback from the team. I've added that to my todo.

**Comment by scordio** [comment]
@snicoll I'm also not that happy with it, maybe duplication of tests might be enough here. I am open to any better proposals. Thanks for taking a look!

**Comment by scordio** [comment]
I deleted my previous comment, I mixed up different things. Thanks for looking into it!

**Comment by scordio** [comment]
Out of curiosity, what concerns do you have about subclassing? I had the assumption that Jpa is a prerequisite to use `spring-data-envers`, therefore any configuration applied by the `JpaRepositoriesRegistrar` should be also applied with the new one.

**Comment by wilkinsona** [comment]
I have just discovered that there's no reference documentation for Spring Data Envers. Unfortunately we can't afford the support burden of an undocumented project so I'm marking this one as blocked.

**Comment by scordio** [comment]
Got it, thanks for looking into it. I'll get in touch with the project to understand if they are willing to improve the docs and accept some help.

**Comment by scordio** [comment]
Hello team, @schauder has taken care of the Spring Data Envers reference documentation. Is that enough to unblock this PR?

**Comment by scordio** [comment]
spring-projects/spring-data-envers#289 might influence this PR.

**Comment by scordio** [comment]
I updated the PR to use the composed annotation introduced with spring-projects/spring-data-envers#290.

In case you have any suggestions on how to avoid the subclass, I'm happy to work in that direction.

**Review by snicoll — APPROVED** [review]

**Comment by snicoll** [comment]
@scordio thank you for making your first contribution to Spring Boot.

**Pull request closed without merging**